File: db_cleaner.py
import sys
import os
from google.cloud import firestore

# Force UTF-8 encoding for standard output on Windows
if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8')

CREDENTIALS_PATH = "Backend/MathGo.Api/firebase-credentials.json"

def main():
    if not os.path.exists(CREDENTIALS_PATH):
        print(f"Error: Credentials file not found at {CREDENTIALS_PATH}")
        sys.exit(1)

    print("Connecting to Firestore using credentials...")
    db = firestore.Client.from_service_account_json(CREDENTIALS_PATH)
    
    # Check execution mode
    clean_mode = "--clean" in sys.argv
    print(f"Mode: {'CLEANUP & RESTRUCTURING (Write)' if clean_mode else 'ANALYSIS (Read-Only)'}")
    print("-" * 50)

    # 1. Analyze / Clean Users collection
    print("\n--- Collection: users ---")
    users_ref = db.collection("users")
    docs = users_ref.stream()

    obsolete_fields = ["totalXp", "totalXP", "gems", "streak", "gamification", "level", "updatedAt"]
    
    total_users = 0
    users_with_obsolete = 0
    clean_users = 0
    
    users_report = []
    
    for doc in docs:
        total_users += 1
        data = doc.to_dict()
        uid = doc.id
        email = data.get("email", "No Email")
        name = data.get("name", "No Name")
        
        found_obsolete = [f for f in obsolete_fields if f in data]
        
        if found_obsolete:
            users_with_obsolete += 1
            users_report.append({
                "uid": uid,
                "email": email,
                "name": name,
                "fields": list(data.keys()),
                "obsolete": found_obsolete,
            })
            
            if clean_mode:
                doc_ref = users_ref.document(uid)
                update_dict = {f: firestore.DELETE_FIELD for f in found_obsolete}
                doc_ref.update(update_dict)
                print(f"🧹 Cleaned user {uid} ({email}): Removed {found_obsolete}")
        else:
            clean_users += 1
            users_report.append({
                "uid": uid,
                "email": email,
                "name": name,
                "fields": list(data.keys()),
                "obsolete": [],
            })

    # 2. Analyze / Add weeklyActivity & weeklyXp in mathgo_progress collection
    print("\n--- Collection: mathgo_progress ---")
    progress_ref = db.collection("mathgo_progress")
    progress_docs = progress_ref.stream()
    
    total_progress = 0
    progress_missing_weekly = 0
    
    progress_report = []
    
    for doc in progress_docs:
        total_progress += 1
        data = doc.to_dict()
        uid = doc.id
        
        has_weekly = "weeklyActivity" in data
        has_weekly_xp = "weeklyXp" in data
        
        weekly_activity = data.get("weeklyActivity", [0, 0, 0, 0, 0, 0, 0])
        if not isinstance(weekly_activity, list) or len(weekly_activity) != 7:
            weekly_activity = [0, 0, 0, 0, 0, 0, 0]
            has_weekly = False
            
        calculated_weekly_xp = sum(weekly_activity)
        current_weekly_xp = data.get("weeklyXp")
        
        needs_update = not has_weekly or not has_weekly_xp or current_weekly_xp != calculated_weekly_xp
        
        if needs_update:
            progress_missing_weekly += 1
            progress_report.append({
                "uid": uid,
                "has_weekly": has_weekly,
                "has_weekly_xp": has_weekly_xp,
                "calculated_weekly_xp": calculated_weekly_xp
            })
            
            if clean_mode:
                doc_ref = progress_ref.document(uid)
                doc_ref.update({
                    "weeklyActivity": weekly_activity,
                    "weeklyXp": calculated_weekly_xp
                })
                print(f"📈 Restructured progress {uid}: Set weeklyActivity and weeklyXp={calculated_weekly_xp}")
        else:
            progress_report.append({
                "uid": uid,
                "has_weekly": True,
                "has_weekly_xp": True,
                "calculated_weekly_xp": calculated_weekly_xp
            })

    print("\n" + "=" * 50)
    print("RESTRUCTURING SUMMARY REPORT")
    print("=" * 50)
    print(f"Collection 'users':")
    print(f"  Total user documents: {total_users}")
    print(f"  Documents with obsolete fields: {users_with_obsolete}")
    print(f"  Clean documents: {clean_users}")
    print(f"\nCollection 'mathgo_progress':")
    print(f"  Total progress documents: {total_progress}")
    print(f"  Documents missing/outdated 'weeklyActivity' or 'weeklyXp': {progress_missing_weekly}")
    print("=" * 50)
    
    print("\nDetailed Users Report:")
    for item in users_report:
        status_icon = "✅" if not item["obsolete"] else "❌"
        print(f"{status_icon} User {item['uid']} ({item['name']} - {item['email']})")
        print(f"   All Fields: {item['fields']}")
        if item["obsolete"]:
            print(f"   Obsolete fields detected: {item['obsolete']}")

    print("\nDetailed Progress Report:")
    for item in progress_report:
        status_icon = "✅" if (item.get("has_weekly", True) and item.get("has_weekly_xp", True)) else "⚠️ (Needs update)"
        print(f"{status_icon} Progress for UID: {item['uid']} (Calculated weeklyXp={item['calculated_weekly_xp']})")

    if not clean_mode:
        if users_with_obsolete > 0 or progress_missing_weekly > 0:
            print("\n👉 To run the database restructuring and clean obsolete fields, run:")
            print("   python db_cleaner.py --clean")
        else:
            print("\n🎉 Database is fully clean and restructured!")

if __name__ == "__main__":
    main()

File: db_cleaner.js
const { initializeApp, cert } = require("firebase-admin/app");
const { getFirestore, FieldValue } = require("firebase-admin/firestore");
const fs = require("fs");
const path = require("path");

const CREDENTIALS_PATH = path.join(__dirname, "Backend/MathGo.Api/firebase-credentials.json");

if (!fs.existsSync(CREDENTIALS_PATH)) {
    console.error(`Error: Credentials file not found at ${CREDENTIALS_PATH}`);
    process.exit(1);
}

const serviceAccount = require(CREDENTIALS_PATH);

initializeApp({
    credential: cert(serviceAccount)
});

const db = getFirestore();

const cleanMode = process.argv.includes("--clean");
console.log(`Mode: ${cleanMode ? 'CLEANUP & RESTRUCTURING (Write)' : 'ANALYSIS (Read-Only)'}`);
console.log("-".repeat(50));

async function run() {
    try {
        // 1. users collection
        console.log("\n--- Collection: users ---");
        const usersSnapshot = await db.collection("users").get();
        const obsoleteFields = ["totalXp", "totalXP", "gems", "streak", "gamification", "level", "updatedAt"];
        
        let totalUsers = 0;
        let usersWithObsolete = 0;
        let cleanUsers = 0;
        
        const usersReport = [];
        
        for (const doc of usersSnapshot.docs) {
            totalUsers++;
            const data = doc.data();
            const uid = doc.id;
            const email = data.email || "No Email";
            const name = data.name || "No Name";
            
            const foundObsolete = obsoleteFields.filter(f => f in data);
            
            if (foundObsolete.length > 0) {
                usersWithObsolete++;
                usersReport.push({
                    uid,
                    email,
                    name,
                    fields: Object.keys(data),
                    obsolete: foundObsolete
                });
                
                if (cleanMode) {
                    const updateDict = {};
                    foundObsolete.forEach(f => {
                        updateDict[f] = FieldValue.delete();
                    });
                    await db.collection("users").doc(uid).update(updateDict);
                    console.log(`🧹 Cleaned user ${uid} (${email}): Removed [${foundObsolete.join(", ")}]`);
                }
            } else {
                cleanUsers++;
                usersReport.push({
                    uid,
                    email,
                    name,
                    fields: Object.keys(data),
                    obsolete: []
                });
            }
        }
        
        // 2. mathgo_progress collection
        console.log("\n--- Collection: mathgo_progress ---");
        const progressSnapshot = await db.collection("mathgo_progress").get();
        
        let totalProgress = 0;
        let progressMissingWeekly = 0;
        const progressReport = [];
        
        for (const doc of progressSnapshot.docs) {
            totalProgress++;
            const data = doc.data();
            const uid = doc.id;
            
            const hasWeekly = "weeklyActivity" in data;
            const hasWeeklyXp = "weeklyXp" in data;
            
            let weeklyActivity = data.weeklyActivity;
            let currentHasWeekly = hasWeekly;
            if (!weeklyActivity || !Array.isArray(weeklyActivity) || weeklyActivity.length !== 7) {
                weeklyActivity = [0, 0, 0, 0, 0, 0, 0];
                currentHasWeekly = false;
            }
            
            const calculatedWeeklyXp = weeklyActivity.reduce((a, b) => a + b, 0);
            const currentWeeklyXp = data.weeklyXp;
            
            const needsUpdate = !currentHasWeekly || !hasWeeklyXp || currentWeeklyXp !== calculatedWeeklyXp;
            
            if (needsUpdate) {
                progressMissingWeekly++;
                progressReport.push({
                    uid,
                    hasWeekly: currentHasWeekly,
                    hasWeeklyXp,
                    calculatedWeeklyXp
                });
                
                if (cleanMode) {
                    await db.collection("mathgo_progress").doc(uid).update({
                        weeklyActivity,
                        weeklyXp: calculatedWeeklyXp
                    });
                    console.log(`📈 Restructured progress ${uid}: Set weeklyActivity and weeklyXp=${calculatedWeeklyXp}`);
                }
            } else {
                progressReport.push({
                    uid,
                    hasWeekly: true,
                    hasWeeklyXp: true,
                    calculatedWeeklyXp
                });
            }
        }
        
        console.log("\n" + "=".repeat(50));
        console.log("RESTRUCTURING SUMMARY REPORT");
        console.log("=".repeat(50));
        console.log("Collection 'users':");
        console.log(`  Total user documents: ${totalUsers}`);
        console.log(`  Documents with obsolete fields: ${usersWithObsolete}`);
        console.log(`  Clean documents: ${cleanUsers}`);
        console.log("\nCollection 'mathgo_progress':");
        console.log(`  Total progress documents: ${totalProgress}`);
        console.log(`  Documents missing/outdated 'weeklyActivity' or 'weeklyXp': ${progressMissingWeekly}`);
        console.log("=".repeat(50));
        
        console.log("\nDetailed Users Report:");
        for (const item of usersReport) {
            const statusIcon = item.obsolete.length === 0 ? "✅" : "❌";
            console.log(`${statusIcon} User ${item.uid} (${item.name} - ${item.email})`);
            console.log(`   All Fields: ${item.fields.join(", ")}`);
            if (item.obsolete.length > 0) {
                console.log(`   Obsolete fields detected: ${item.obsolete.join(", ")}`);
            }
        }
        
        console.log("\nDetailed Progress Report:");
        for (const item of progressReport) {
            const statusIcon = (item.hasWeekly && item.hasWeeklyXp) ? "✅" : "⚠️ (Needs update)";
            console.log(`${statusIcon} Progress for UID: ${item.uid} (Calculated weeklyXp=${item.calculatedWeeklyXp})`);
        }
        
        if (!cleanMode) {
            if (usersWithObsolete > 0 || progressMissingWeekly > 0) {
                console.log("\n👉 To run the database restructuring and clean obsolete fields, run:");
                console.log("   node db_cleaner.js --clean");
            } else {
                console.log("\n🎉 Database is fully clean and restructured!");
            }
        }
        
    } catch (err) {
        console.error("Error during run:", err);
    }
}

run();
